### DISCUSSION TITLE
"Nice" JSONField support

### DISCUSSION THREAD

**Discussion by sowinski** [wagtail/wagtail#8999] [Ideas]
Hi,

I use Wagtail for my multilingual website which works great. I can write nicely content for my blog / news etc.

But there is one feature missing for me. I have a multilingual website with different domains, like
example.com
example.nl
example.fr
example.de
...

And each of them has a pricing page for example.
I started to have a static template for my pricing pages and translated everything with the i18n feature using .po files which ships with Django. But this is not nice if you want to change the translation/content often.
So I decided to build a wagtail page and each text element in the template is replaced by a "model.field". So from here I can start editing/translating the page in wagtail.
But changes in the templates always required changes in the database.

So I am looking for a more flexible way. I thought it would be a good idea to use a JSONField. With a JSONField I can add new "fields" without changing the database layout. My translators/editors have struggle to work with JSON. They sometimes break the JSON syntax and can't save the changes.
So my idea would be to implement a nice JSON Widget which can help everyone to write and edit JSON.

How would you do this? What do you think about this?

**Comment by gasman** [comment]
(have moved this to a discussion, as it's probably too open-ended to handle as an issue, and unclear whether this is best handled in Wagtail itself or an add-on package)

**Reply by sowinski** [reply]
Okay thank you

> and unclear whether this is best handled in Wagtail itself or an add-on package

How would you handle this case?

**Comment by ababic** [comment]
@sowinski It's possible I'm not quite understanding the scale of problem, but I would say a dedicated field for each piece of data is the best way to go, because:
1. The field labels can be translated, so the editor sees an interface that make sense to them (in their preferred language)
2. If you rename a model field, the existing value will be retained, and for more complex changes, you have the option to write a Django data migration to populate new field values from old ones before removing the old fields.
3. If you want placement on the page to be flexible, you could add a simple StaticBlock option to a StreamField, and have that pull out/display the model field values in the position chosen by the editor.

I would also consider the following options before having editors edit raw JSON in the page editor:

1. Hardcoding the price information in the page template AFTER the editable content.

Or, if the price information needs to appear in different places on each version of the page:

1. Write a custom StaticBlock to render a template with the pricing bits hardcoded. 
2. Write a custom StaticBlock that pulls in pricing data from another model, allowing the data to be dynamic without becoming part of the page's revision history.

**Reply by sowinski** [reply]
I have like 100 fields. Using fields have a few downsides:
1) The editing page becomes incredible long, which feels confusing to me
2) Everytime I change my template and fields I have to change the Django code and make migrations
(Enhancement: It would be nice, if there could be a "fast editing" view (Like a table) to decrease the size of the page and make it easier to overlook more fields at once)

I ended up using this JSON widget: https://github.com/cursive-works/wagtail-json-widget
Now I have a nice editor (You can change the editor to a form view, so the editor don't need to write any JSON code anymore) and in my Page there is a dynamic selector for a template. The editor will see a basic/default template in English and can adopt this to a different language. 
With this solution I can easily write new templates, modify them and upload them without touching the python code. I also do not need to restart my webserver or update my database with new changes.

**Reply by ababic** [reply]
> The editing page becomes incredible long, which feels confusing to me

Wrapping multiple related panels in a [MultiFieldPanel](https://docs.wagtail.org/en/stable/reference/pages/panels.html#multifieldpanel) can help a great deal with this. Occasionally I will also use [FieldRowPanel](https://docs.wagtail.org/en/stable/reference/pages/panels.html#fieldrowpanel) to make better use of horizontal space (which I think is similar to your Table layout idea). NOTE: The `classname` argument (available on most panels) is really handy for adding a touch of custom styling where needed.

> Everytime I change my template and fields I have to change the Django code and make migrations

But what happens when you change the JSON structure expected by the template? What do you do about the existing values in that scenario? At least with model field changes you have a rock-solid framework (field and data migrations) for helping you convert existing data. If there are significant changes being made to the page, there's work to do either way, right?

**Comment by ababic** [comment]
With all of the above said, if you really REALLY want a field that stores JSON and has a nice widget to facilitate editing, find a widget that works for you, and use FieldPanel's [`widget` option](https://docs.wagtail.org/en/stable/reference/pages/panels.html#wagtail.admin.panels.FieldPanel.widget) to use it in the form. 

An option I have used successfully before is: https://pypi.org/project/django-json-widget/
